File: Cpp/0150_evaluate_reverse_polish_notation.cpp
// https://leetcode.cn/problems/evaluate-reverse-polish-notation/

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        vector<int> nums;
        for (string token: tokens) {
            if ((token.size() > 1) || ((token[0] >= '0') && (token[0] <= '9'))) {
                int num = stoi(token);
                nums.push_back(num);
            }
            else {
                int second = nums.back();
                nums.pop_back();
                int first = nums.back();
                nums.pop_back();
                int num;
                switch(token[0]) {
                    case '+': {
                        num = first + second;
                        break;
                    }
                    case '-': {
                        num = first - second;
                        break;
                    }
                    case '*': {
                        num = first * second;
                        break;
                    }
                    default: {
                        num = first / second;
                        break;
                    }
                }
                nums.push_back(num);
            }
        }
        return nums[0];
    }
};

File: Python3/0150_evaluate_reverse_polish_notation.py
# https://leetcode.cn/problems/evaluate-reverse-polish-notation/

class Solution:
    def evalRPN(self, tokens):
        """
        Inputs:
            tokens: list[str]
        Outputs:
            res: int
        """

        nums = []
        ops = ["+", "-", "*", "/"]
        for token in tokens:
            if token not in ops:
                num = int(token)
                nums.append(num)
            else:
                second = nums.pop()
                first = nums.pop()
                if token == "+":
                    tmp = first + second
                elif token == "-":
                    tmp = first - second
                elif token == "*":
                    tmp = first * second
                else:
                    tmp = int(first / second)
                nums.append(tmp)

        return nums[0]

solu = Solution()
tokens = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]
print(solu.evalRPN(tokens))